exercises/template/ex02_oo_basics/src/shape.hpp

```diff
@@ -0,0 +1,85 @@
+#ifndef SHAPE_HPP
+
+#define SHAPE_HPP
+
+#include <iostream>
+#include <math.h>
+#include <memory>
+#include <vector>
+
+using namespace std;
+
+class Shape {
+public:
+  virtual int sides() const = 0;
+  virtual double perimeter() const = 0;
+  virtual double area() const = 0;
+  virtual string type() const = 0;
+};
+
+ostream &operator<<(ostream &os, const Shape *shape);
+
+class Circle : public Shape {
+private:
+  double radius;
+
+public:
+  explicit Circle(double r) : radius(r) {}
+  int sides() const override;
+  double perimeter() const override;
+  double area() const override;
+  string type() const override;
+};
+
+class Square : public Shape {
+private:
+  double side;
+
+public:
+  explicit Square(double s) : side(s) {}
+  int sides() const override;
+  double perimeter() const override;
+  double area() const override;
+  string type() const override;
+};
+
+class Rectangle : public Shape {
```

Comment:
In general we prefer separate `*.h` `*.cpp` files for classes but as this is an exercise don't bother changing it.